asf_core_data/notebooks/explore_address_matching_sample.py

```diff
@@ -0,0 +1,485 @@
+
+# %%
+rcd.from_predictions(y_true=sample["confident_match"], y_pred=sample["address_score"])
+
+# %%
+youdens_thresholds_country = {}
+for country in ["England", "Scotland", "Wales"]:
+    fpr, tpr, thresholds = roc_curve(
+        y_true=sample[sample["country"] == country]["confident_match"],
+        y_score=sample[sample["country"] == country]["address_score"],
+    )
+    ## Calculate Youden's J statistic
+    ## sensitivity + specificity - 1 = TPR - FPR
+    _df = pd.DataFrame({"youdens": tpr - fpr, "threshold": thresholds}).sort_values(
+        by="youdens", ascending=False
+    )
+    _df["removed_candidates"] = [
+        (sample["address_score"] < threshold).sum() for threshold in _df["threshold"]
+    ]
+    _df["removed_true_matches"] = [
+        (sample[sample["confident_match"] == 1]["address_score"] < threshold).sum()
+        for threshold in _df["threshold"]
+    ]
+    youdens_thresholds_country[country] = _df
+
+youdens_thresholds_country
+
+# %%
+youdens_thresholds_numeric = {}
+for _, key in zip([0, 1], ["non_numeric", "numeric"]):
+    fpr, tpr, thresholds = roc_curve(
+        y_true=sample[sample["numeric_address_component"] == _]["confident_match"],
+        y_score=sample[sample["numeric_address_component"] == _]["address_score"],
+    )
+    ## Calculate Youden's J statistic
+    ## sensitivity + specificity - 1 = TPR - FPR
+    _df = pd.DataFrame({"youdens": tpr - fpr, "threshold": thresholds}).sort_values(
+        by="youdens", ascending=False
+    )
+    _df["removed_candidates"] = [
+        (sample["address_score"] < threshold).sum() for threshold in _df["threshold"]
+    ]
+    _df["removed_true_matches"] = [
+        (sample[sample["confident_match"] == 1]["address_score"] < threshold).sum()
+        for threshold in _df["threshold"]
+    ]
+    youdens_thresholds_numeric[key] = _df
+
+youdens_thresholds_numeric
+
+# %% [markdown]
+# Even when we search for optimum threshold by country or by numeric/alpha address, it looks like there could be a potential optimum address score threshold between 0.84-0.869231.
+#
+# This optimum is based on the following assumptions:
+#
+# - We want to maximise Youden's J-statistic
+# - i.e. sensitivity and specificity are equally important: removing false positives is as important as retaining true positives
+#
+# Based on these results and the confident match rate by address scores assessed above, we should consider increasing the address score threshold at least to 0.74 and potentially to somewhere between 0.77 and 0.87.
+#
+# However, we lose true positive matches with every increase in the threshold. Also, simply increasing the threshold cannot remove many of the erroneous matches in non-numeric addresses because many of these have very high similarity scores. We will need a different methodology to address these errors. We will explore alternative improvements to address matching in a separate analysis.
```

Comment:
I guess a question I have is: is the best we can get from this model (by doing the above improvements) enough? it's certainly better than we have now, but we potentially need a better address matching model.